README.md

```diff
@@ -50,17 +50,74 @@ engine.push(function(req, res, next, end){
 })
 ```
 
-RpcEngines can be nested by converting them to middleware `asMiddleware(engine)`
+RpcEngines can be nested by converting them to middleware using `asMiddleware(engine)`:
 
 ```js
-const asMiddleware = require('json-rpc-engine/lib/asMiddleware')
+const asMiddleware = require('json-rpc-engine/src/asMiddleware')
 
 let engine = new RpcEngine()
 let subengine = new RpcEngine()
 engine.push(asMiddleware(subengine))
 ```
 
-### gotchas
+### Asynchronous Middlewares
+
+If you require your middleware function to be asynchronous, use `createAsyncMiddleware`:
+
+```js
+const createAsyncMiddleware = require('json-rpc-engine/src/createAsyncMiddleware')
+
+let engine = new RpcEngine()
+engine.push(createAsyncMiddleware(async (req, res, next) => {
+  req.result = 42
+  next()
+}))
+```
+
+Asynchronous middlewares do not take an `end` callback.
+Instead, the request ends if the middleware returns without calling `next()`:
+
+```js
+engine.push(createAsyncMiddleware(async (req, res, next) => {
+  req.result = 42
+  /* The request will end when this returns */
+}))
+```
+
+The `next` callback of asynchronous middlewares also don't take return handlers.
+Instead, you can `await next()`.
+When the execution of the middleware resumes, you can work with the response.
+
+```js
+engine.push(createAsyncMiddleware(async (req, res, next) => {
+  req.result = 42
+  await next()
+  /* Now you can work with the response */
+  addToMetrics(res)
+}))
+```
+
+You can freely mix synchronous and asynchronous middlewares:
```

Comment:
Maybe this is a better term to use than "synchronous"?
```suggestion
You can freely mix callback-based and asynchronous middlewares:
```

Reply:
2561ecf